### ISSUE TITLE
VersionNumber replacement

### ISSUE THREAD

**Issue by xenoterracide** [gradle/gradle#28755]
### Issue type

Missing information

### Problem description

As a plugin developer, what can I use instead of this to ensure that my versions compare correctly according to gradle's own rules? 

note: version generation is pretty popular when you look at the plugin portal, getting it right isn't super easy either.

### Page with the problem

https://docs.gradle.org/current/javadoc/org/gradle/util/VersionNumber.html

**Comment by ov7a** [comment]
The issue is in the backlog of the relevant team and is prioritized by them.

***

Related:
* https://github.com/gradle/gradle/issues/28515

**Comment by mhsmith** [comment]
In #17450, @big-guy said the following:

> The problem with `VersionNumber` is that there is no single standard way to format a version. This class takes an approach that's _like_ many version formats, but it was always limited to what was useful for our specific needs and wasn't general purpose.
> 
> It's very tempting to pass a String into this and get something out that's a "version", but there are no guarantees that it makes any sense unless you know how the version format should be interpreted. The thinking is that it's best to leave version parsing/interpretation up to the plugin that has the context about the version.

But in the context of a Gradle build script or plugin, the only interpretation you're likely to want is the same interpretation as Gradle itself. And the only reliable way to achieve that is for the VersionNumber API to remain public.

**Comment by xenoterracide** [comment]
Yes this API should remain public although there are implementation issues in it that should be corrected. That or the internal API that would actually provide a correct sort should be exposed.

**Comment by mhsmith** [comment]
What implementation issues do you mean, and do they already have issues in the GitHub tracker?

**Comment by xenoterracide** [comment]
> What implementation issues do you mean, and do they already have issues in the GitHub tracker?

Yes they do. I opened an issue here somewhere... I don't feel like tracking it down right now. However what version number considers to be the latest is not the same as what Gradle will actually download as the latest given the same set of version strings.

**Comment by mhsmith** [comment]
I  guess you mean this one:
* #28805

**Comment by xenoterracide** [comment]
I suggest using mavens stuff for the time being. I can find that code in my semver plugin if you want. Basically it's use 3rd party tools and hope for the best right now.

**Comment by mhsmith** [comment]
> But in the context of a Gradle build script or plugin, the only interpretation you're likely to want is the same interpretation as Gradle itself. And the only reliable way to achieve that is for the VersionNumber API to remain public.

Actually, now that I look at the [VersionNumber implementation](https://github.com/gradle/gradle/blob/v8.14.3/subprojects/core/src/main/java/org/gradle/util/VersionNumber.java), it doesn't implement Gradle's full [version ordering algorithm](https://docs.gradle.org/current/userguide/dependency_versions.html#sec:version-ordering) at all, but something much more limited.

[`ModuleDescriptor.Version`](https://docs.oracle.com/javase/9/docs/api/java/lang/module/ModuleDescriptor.Version.html) isn't a perfect match either, but it's much closer, and it's in the Java standard library. See examples at https://stackoverflow.com/a/65077344.

Related:
* https://github.com/gradle/gradle/issues/34546
